File: ifw-api/src/main/java/com/merxury/ifw/IntentFirewallImpl.java
package com.merxury.ifw;

import android.content.Context;

import androidx.annotation.NonNull;

import com.elvishew.xlog.Logger;
import com.elvishew.xlog.XLog;
import com.merxury.ifw.entity.Activity;
import com.merxury.ifw.entity.Broadcast;
import com.merxury.ifw.entity.Component;
import com.merxury.ifw.entity.ComponentFilter;
import com.merxury.ifw.entity.ComponentType;
import com.merxury.ifw.entity.Rules;
import com.merxury.ifw.entity.Service;
import com.merxury.libkit.RootCommand;
import com.merxury.libkit.utils.FileUtils;
import com.merxury.libkit.utils.StorageUtils;

import org.simpleframework.xml.Serializer;
import org.simpleframework.xml.core.Persister;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class IntentFirewallImpl implements IntentFirewall {

    private static final String EXTENSION = ".xml";
    private static final String FILTER_TEMPLATE = "%s/%s";
    private static IntentFirewallImpl instance;
    private final Logger logger = XLog.tag("IntentFirewallImpl").build();
    private final String filename;
    private Rules rules;
    private final String tmpPath;
    private final String destPath;
    private final String packageName;
    private final String cacheDir;

    public IntentFirewallImpl(Context context, String packageName) {
        this.packageName = packageName;
        this.filename = packageName + EXTENSION;
        cacheDir = context.getCacheDir().toString() + File.separator;
        tmpPath = cacheDir + filename;
        destPath = StorageUtils.getIfwFolder() + filename;
        openFile();
    }

    public static IntentFirewall getInstance(@NonNull Context context, String packageName) {
        synchronized (IntentFirewallImpl.class) {
            if (instance == null || !instance.getPackageName().equals(packageName)) {
                instance = new IntentFirewallImpl(context, packageName);
            }
        }
        return instance;
    }

    public Rules getRules() {
        return rules;
    }

    @Override
    public void save() throws Exception {
        ensureNoEmptyTag();
        if (rules.getActivity() == null && rules.getBroadcast() == null && rules.getService() == null) {
            // If there is no rules presented, delete rule file (if exists)
            FileUtils.delete(destPath, true);
            return;
        }
        File file = new File(tmpPath);
        Serializer serializer = new Persister();
        if (file.exists()) {
            if (!file.delete()) {
                throw new IOException("Cannot delete file: " + tmpPath);
            }
        }
        serializer.write(rules, file);
        FileUtils.cat(tmpPath, destPath);
        FileUtils.chmod(destPath, 644, false);
    }

    @Override
    public boolean add(String packageName, String componentName, ComponentType type) {
        boolean result = false;
        switch (type) {
            case ACTIVITY:
                if (rules.getActivity() == null) {
                    rules.setActivity(new Activity());
                }
                result = addComponentFilter(packageName, componentName, rules.getActivity());
                break;
            case BROADCAST:
                if (rules.getBroadcast() == null) {
                    rules.setBroadcast(new Broadcast());
                }
                result = addComponentFilter(packageName, componentName, rules.getBroadcast());
                break;
            case SERVICE:
                if (rules.getService() == null) {
                    rules.setService(new Service());
                }
                result = addComponentFilter(packageName, componentName, rules.getService());
                break;
            default:
                break;
        }
        return result;
    }

    private boolean addComponentFilter(String packageName, String componentName, Component component) {
        if (component == null) {
            return false;
        }
        List<ComponentFilter> filters = component.getComponentFilters();
        if (filters == null) {
            filters = new ArrayList<>();
            component.setComponentFilters(filters);
        }
        String filterRule = formatName(packageName, componentName);
        //Duplicate filter detection
        for (ComponentFilter filter : filters) {
            if (filter.getName().equals(filterRule)) {
                return false;
            }
        }
        filters.add(new ComponentFilter(filterRule));
        logger.i("Added component:" + packageName + "/" + componentName);
        return true;
    }

    private boolean removeComponentFilter(String packageName, String componentName, Component component) {
        if (component == null) {
            return false;
        }
        List<ComponentFilter> filters = component.getComponentFilters();
        if (filters == null) {
            filters = new ArrayList<>();
        }
        String filterRule = formatName(packageName, componentName);
        for (ComponentFilter filter : new ArrayList<>(filters)) {
            if (filterRule.equals(filter.getName())) {
                filters.remove(filter);
            }
        }
        return true;
    }

    @Override
    public boolean remove(String packageName, String componentName, ComponentType type) {
        boolean result = false;
        switch (type) {
            case ACTIVITY:
                result = removeComponentFilter(packageName, componentName, rules.getActivity());
                break;
            case BROADCAST:
                result = removeComponentFilter(packageName, componentName, rules.getBroadcast());
                break;
            case SERVICE:
                result = removeComponentFilter(packageName, componentName, rules.getService());
                break;
            default:
                break;
        }
        return result;
    }

    @Override
    public boolean getComponentEnableState(String packageName, String componentName) {
        List<ComponentFilter> filters = new ArrayList<>();
        if (rules.getActivity() != null) {
            filters.addAll(rules.getActivity().getComponentFilters());
        }
        if (rules.getBroadcast() != null) {
            filters.addAll(rules.getBroadcast().getComponentFilters());
        }
        if (rules.getService() != null) {
            filters.addAll(rules.getService().getComponentFilters());
        }
        return getFilterEnableState(packageName, componentName, filters);
    }

    @Override
    public void clear() {
        clear(filename);
    }

    @Override
    public void clear(String name) {
        if (name == null) {
            return;
        }
        String rulePath = StorageUtils.getIfwFolder() + filename;
        logger.d("delete file: " + rulePath);
        if (!rulePath.contains("ifw")) {
            logger.e("Something wrong happened, folder is not correct");
            return;
        }
        RootCommand.runBlockingCommand("rm -rf " + rulePath);
    }

    @Override
    public void reload() {
        openFile();
    }

    private boolean getFilterEnableState(String packageName, String componentName, List<ComponentFilter> componentFilters) {
        if (componentFilters == null) {
            return true;
        }
        for (ComponentFilter filter : componentFilters) {
            if (filter == null) {
                return false;
            }
            String filterName = formatName(packageName, componentName);
            if (filterName.equals(filter.getName())) {
                return false;
            }
        }
        return true;
    }

    private void openFile() {
        removeCache();
        if (FileUtils.isExist(destPath)) {
            String ruleContent = FileUtils.read(destPath);
            Serializer serializer = new Persister();
            try {
                File file = new File(cacheDir, filename);
                FileWriter writer = new FileWriter(file);
                writer.write(ruleContent);
                writer.close();
                rules = serializer.read(Rules.class, file);
            } catch (Exception e) {
                handleException(e);
                rules = new Rules();
            }
        } else {
            rules = new Rules();
        }
    }

    public boolean removeCache() {
        try {
            File cacheFile = new File(cacheDir, filename);
            boolean result = cacheFile.exists() && cacheFile.delete();
            if (result) {
                reload();
            }
            return result;
        }catch (Exception e) {
            logger.e("Cannot delete cache file: " + cacheDir, e);
            return false;
        }
    }

    private void ensureNoEmptyTag() {
        if (rules.getActivity() != null && rules.getActivity().getComponentFilters() != null) {
            if (rules.getActivity().getComponentFilters().size() == 0) {
                rules.setActivity(null);
            }
        }
        if (rules.getBroadcast() != null && rules.getBroadcast().getComponentFilters() != null) {
            if (rules.getBroadcast().getComponentFilters().size() == 0) {
                rules.setBroadcast(null);
            }
        }
        if (rules.getService() != null && rules.getService().getComponentFilters() != null) {
            if (rules.getService().getComponentFilters().size() == 0) {
                rules.setService(null);
            }
        }
    }

    private String formatName(String packageName, String name) {
        return String.format(FILTER_TEMPLATE, packageName, name);
    }

    private void handleException(Exception e) {
        logger.e(e.getMessage());
        e.printStackTrace();
    }

    @Override
    public String getPackageName() {
        return packageName;
    }
}

File: core/ifw-api/src/main/java/com/merxury/ifw/IntentFirewallImpl.kt
/*
 * Copyright 2022 Blocker
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       https://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

package com.merxury.ifw

import com.elvishew.xlog.XLog
import com.merxury.blocker.core.utils.FileUtils
import com.merxury.blocker.core.utils.PermissionUtils
import com.merxury.blocker.core.utils.StorageUtils
import com.merxury.ifw.entity.Activity
import com.merxury.ifw.entity.Broadcast
import com.merxury.ifw.entity.Component
import com.merxury.ifw.entity.ComponentFilter
import com.merxury.ifw.entity.ComponentType
import com.merxury.ifw.entity.Rules
import com.merxury.ifw.entity.Service
import com.merxury.ifw.exception.RootUnavailableException
import com.topjohnwu.superuser.io.SuFile
import com.topjohnwu.superuser.io.SuFileInputStream
import com.topjohnwu.superuser.io.SuFileOutputStream
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.simpleframework.xml.Serializer
import org.simpleframework.xml.core.Persister

class IntentFirewallImpl(override val packageName: String) : IntentFirewall {

    private val logger = XLog.tag("IntentFirewallImpl").build()
    private val filename: String = "$packageName$EXTENSION"
    private val destFile = SuFile(StorageUtils.getIfwFolder() + filename)
    private var rule: Rules = Rules()

    override suspend fun load() = withContext(Dispatchers.IO) {
        if (PermissionUtils.isRootAvailable() && destFile.exists()) {
            val serializer: Serializer = Persister()
            try {
                val input = SuFileInputStream.open(destFile)
                rule = serializer.read(Rules::class.java, input)
            } catch (e: Exception) {
                logger.e("Error reading rules file $destFile:", e)
            }
        }
        return@withContext this@IntentFirewallImpl
    }

    override suspend fun save() {
        withContext(Dispatchers.IO) {
            if (!PermissionUtils.isRootAvailable()) {
                throw RootUnavailableException()
            }
            ensureNoEmptyTag()
            if (rule.activity == null && rule.broadcast == null && rule.service == null) {
                // If there is no rules presented, delete rule file (if exists)
                clear()
                return@withContext
            }
            SuFileOutputStream.open(destFile).use {
                val serializer: Serializer = Persister()
                serializer.write(rule, it)
            }
            FileUtils.chmod(destFile.absolutePath, 644, false)
            logger.i("Saved $destFile")
        }
    }

    override suspend fun clear() {
        withContext(Dispatchers.IO) {
            if (!PermissionUtils.isRootAvailable()) {
                throw RootUnavailableException()
            }
            logger.d("Clear IFW rule $filename")
            if (destFile.exists()) {
                destFile.delete()
            }
            rule = Rules()
        }
    }

    override suspend fun add(
        packageName: String,
        componentName: String,
        type: ComponentType?
    ): Boolean {
        if (!PermissionUtils.isRootAvailable()) {
            logger.e("Root unavailable, cannot add rule")
            throw RootUnavailableException()
        }
        var result = false
        when (type) {
            ComponentType.ACTIVITY -> {
                if (rule.activity == null) {
                    rule.activity = Activity()
                }
                result = addComponentFilter(packageName, componentName, rule.activity)
            }

            ComponentType.BROADCAST -> {
                if (rule.broadcast == null) {
                    rule.broadcast = Broadcast()
                }
                result = addComponentFilter(packageName, componentName, rule.broadcast)
            }

            ComponentType.SERVICE -> {
                if (rule.service == null) {
                    rule.service = Service()
                }
                result = addComponentFilter(packageName, componentName, rule.service)
            }

            else -> {}
        }
        return result
    }

    override suspend fun remove(
        packageName: String,
        componentName: String,
        type: ComponentType?
    ): Boolean {
        if (!PermissionUtils.isRootAvailable()) {
            logger.e("Root unavailable, cannot remove rule")
            throw RootUnavailableException()
        }
        return when (type) {
            ComponentType.ACTIVITY -> removeComponentFilter(
                packageName,
                componentName,
                rule.activity
            )

            ComponentType.BROADCAST -> removeComponentFilter(
                packageName,
                componentName,
                rule.broadcast
            )

            ComponentType.SERVICE -> removeComponentFilter(
                packageName,
                componentName,
                rule.service
            )

            else -> false
        }
    }

    override suspend fun getComponentEnableState(
        packageName: String,
        componentName: String
    ): Boolean {
        val filters: MutableList<ComponentFilter> = ArrayList()
        rule.activity?.let {
            filters.addAll(it.componentFilters)
        }
        rule.broadcast?.let {
            filters.addAll(it.componentFilters)
        }
        rule.service?.let {
            filters.addAll(it.componentFilters)
        }
        return getFilterEnableState(packageName, componentName, filters)
    }

    private fun ensureNoEmptyTag() {
        if (rule.activity != null && rule.activity.componentFilters.isNullOrEmpty()) {
            rule.activity = null
        }
        if (rule.broadcast != null && rule.broadcast.componentFilters.isNullOrEmpty()) {
            rule.broadcast = null
        }
        if (rule.service != null && rule.service.componentFilters.isNullOrEmpty()) {
            rule.service = null
        }
    }

    private fun addComponentFilter(
        packageName: String,
        componentName: String,
        component: Component?
    ): Boolean {
        if (component == null) {
            return false
        }
        var filters = component.componentFilters
        if (filters == null) {
            filters = ArrayList()
            component.componentFilters = filters
        }
        val filterRule = formatName(packageName, componentName)
        // Duplicate filter detection
        for (filter in filters) {
            if (filter.name == filterRule) {
                return false
            }
        }
        filters.add(ComponentFilter(filterRule))
        logger.i("Added component:$packageName/$componentName")
        return true
    }

    private fun removeComponentFilter(
        packageName: String,
        componentName: String,
        component: Component?
    ): Boolean {
        if (component == null) {
            return false
        }
        var filters = component.componentFilters
        if (filters == null) {
            filters = ArrayList()
        }
        val filterRule = formatName(packageName, componentName)
        for (filter in ArrayList(filters)) {
            if (filterRule == filter.name) {
                filters.remove(filter)
            }
        }
        return true
    }

    private fun getFilterEnableState(
        packageName: String,
        componentName: String,
        componentFilters: List<ComponentFilter>?
    ): Boolean {
        if (componentFilters == null) {
            return true
        }
        for (filter in componentFilters) {
            val filterName = formatName(packageName, componentName)
            if (filterName == filter.name) {
                return false
            }
        }
        return true
    }

    private fun formatName(packageName: String, name: String): String {
        return "$packageName/$name"
    }

    companion object {
        private const val EXTENSION = ".xml"
    }
}
